### PR TITLE
Don't re-authenticate requests which don't have a JWT

### PR DESCRIPTION
**Pull request by brandonSc**
Edge case where this can cause an infinite recursive loop. 

Basically, an authenticated request would 401 (e.g. list secrets but not logged in) which would trigger a login request  (re-auth) -> which would 401 and trigger SSH discovery -> which would request an "auth challenge" -> which would 401, and trigger another login request (re-auth) -> which would trigger SSH discovery again -> ... etc

The solution here prevents the 401 on a login request from retrying.

### CODE REVIEW

File: cloud/client.go [hunk 1]
```diff
@@ -169,6 +169,9 @@ func (c *client) doCall(method, url string, opts ...requestOpt) (int, string, er
 		}
 
 		if status == http.StatusUnauthorized {
+			if !r.hasAuth {
```
Comment (alexcb) [inline cloud/client.go:172]: Should we combine this with the below if? e.g.
```suggestion
			if !r.hasAuth || alreadyReAuthed {
```

it's not clear why `alreadyReAuthed` returns `err` and not `errUnauthorized`; I would have thought they would both return the same error.

**Comment by alexcb** [comment]
It's not completely related to this PR, but I noticed `alreadyReAuthed := false` is set after the initial

```
if r.hasAuth && time.Now().UTC().After(c.authTokenExpiry) {
		if err := c.Authenticate(); err != nil {
			return 0, "", errors.Wrap(err, "failed refreshing expired auth token")
		}
	}
```

we should change this to

```
alreadyReAuthed := false
if r.hasAuth && time.Now().UTC().After(c.authTokenExpiry) {
		if err := c.Authenticate(); err != nil {
			return 0, "", errors.Wrap(err, "failed refreshing expired auth token")
		}
alreadyReAuthed = true
	}
```

File: cloud/client.go (hunk 1 above)
Comment (brandonSc) [reply cloud/client.go:172]: Will give it a try

Comment (brandonSc) [reply cloud/client.go]: Added this

**Review by alexcb — APPROVED** [review]

**Pull request merged**